File: openmole/core/org.openmole.core.batchservicecontrol/src/main/java/org/openmole/core/batchservicecontrol/UsageControl.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.core.batchservicecontrol;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import org.openmole.commons.aspect.eventdispatcher.ObjectModified;

import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.core.model.execution.batch.IAccessToken;
import org.openmole.core.model.execution.batch.IAccessTokenPool;

public class UsageControl implements IUsageControl {

    IAccessTokenPool tokenPool;

    public UsageControl(int nbAccess) {
        super();
        if (nbAccess != Integer.MAX_VALUE) {
            tokenPool = new AccessTokenPool(nbAccess);
        } else {
            tokenPool = new BotomlessTokenPool();
        }
    }

    @Override
    public IAccessToken tryGetToken(long time, TimeUnit unit) throws InterruptedException, TimeoutException {
        return tokenPool.waitAToken(time, unit);
    }

    @Override
    public IAccessToken waitAToken() throws InterruptedException {
        return tokenPool.waitAToken();
    }

    @Override
    public IAccessToken getAccessTokenInterruptly() {
        return tokenPool.getAccessTokenInterruptly();
    }

    @ObjectModified(name = resourceReleased)
    @Override
    public void releaseToken(IAccessToken token) throws InternalProcessingError, UserBadDataError {
        tokenPool.releaseToken(token);
    }

    @Override
    public int getLoad() {
        return tokenPool.getLoad();
    }
}

File: openmole/core/org.openmole.core.batchservicecontrol/src/main/scala/org/openmole/core/batchservicecontrol/UsageControl.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.batchservicecontrol

import org.openmole.commons.aspect.eventdispatcher.ObjectModified

import org.openmole.core.batchservicecontrol.internal.AccessTokenPool
import org.openmole.core.batchservicecontrol.internal.BotomlessTokenPool
import org.openmole.core.model.execution.batch.IAccessToken
import java.util.concurrent.TimeUnit

object UsageControl {
  def apply(nbAccess: Int) = {
    if (nbAccess != Int.MaxValue) {
      new UsageControl(AccessTokenPool(nbAccess));
    } else {
      new UsageControl(BotomlessTokenPool)
    }
  }
}


class UsageControl(tokenPool: IAccessTokenPool) extends IUsageControl {

  override def waitAToken(time: Long, unit: TimeUnit): IAccessToken = tokenPool.waitAToken(time, unit)

  override def waitAToken: IAccessToken = tokenPool.waitAToken

  override def tryGetToken: Option[IAccessToken] = tokenPool.tryGetToken
  
  @ObjectModified(name = IUsageControl.ResourceReleased)
  override def releaseToken(token: IAccessToken) = tokenPool.releaseToken(token)
 
  override def load: Int = tokenPool.load
  
}
